lib/src/package_analysis/lower_bound_constraint_analysis.dart

```diff
@@ -0,0 +1,241 @@
+        await packageAnalysisContext.analysisSession.getResolvedUnit(filePath);
+    if (result is ResolvedUnitResult) {
+      astVisitor.visitCompilationUnit(result.unit);
+    } else {
+      packageAnalysisContext.warning(
+          'Attempting to get a resolved unit resulted in an invalid result.');
+    }
+  }
+
+  return astVisitor.issues.values.whereNotNull().toList();
+}
+
+class _LowerBoundConstraintVisitor extends GeneralizingAstVisitor {
+  /// Maps from [Element.id] to either [LowerBoundConstraintIssue] if the lower
+  /// bound constraint on [LowerBoundConstraintIssue.dependencyPackageName] is
+  /// too low, making [Element] when the lowest allowed version
+  /// [LowerBoundConstraintIssue.lowestVersion] is used, or null otherwise.
+  final Map<int, LowerBoundConstraintIssue?> issues = {};
+
+  /// The name of the package being analyzed. Invocations corresponding to
+  /// definitions within this package will be ignored.
+  final String? rootPackage;
+
+  /// The summaries of each of the dependencies of the target package.
+  final Map<String, PackageShape> dependencySummaries;
+
+  /// The dependencies of the target package, as parsed from its pubspec.
+  final Map<String, HostedDependency> targetDependencies;
+
+  /// The versions of the dependencies of the target package, as installed.
+  final Map<String, Version> dependencyInstalledVersions;
+
+  /// Log a warning that something unexpected happened.
+  final void Function(String message) warning;
+
+  _LowerBoundConstraintVisitor({
+    required this.rootPackage,
+    required this.warning,
+    required this.dependencySummaries,
+    required this.targetDependencies,
+    required this.dependencyInstalledVersions,
+  });
+
+  // TODO: consider FunctionReference, PropertyAccess
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    // an invocation of a top-level function or a class method
+    super.visitMethodInvocation(node);
+
+    final element = node.methodName.staticElement;
+
+    if (element == null) {
+      warning(
+          'Element associated with identifier ${node.methodName.name} could not be resolved.');
+      return;
+    }
+
+    final elementId = element.id;
+    final libraryUri = element.library!.identifier;
+    final symbolName = element.name!;
+    final packageName = packageFromLibraryUri(libraryUri);
+
+    if (issues.containsKey(elementId)) {
+      if (issues[elementId] == null) {
+        // if there is no problem with this element, return immediately
+        return;
+      } else {
+        // we have seen this symbol before
+        // TODO: handle this case by adding a reference to issues[elementId]
+        return;
+      }
+    }
+
+    // if the name of the package can't be resolved,
+    // or it is the same as the name of the target package (or flutter)
+    // then this symbol cannot be analyzed
+    if (packageName == null ||
+        packageName == rootPackage ||
+        packageName == 'flutter') {
```

Comment:
There are other dependencies SDK dependencies than just `flutter´.
Maybe we should explicitly ignore all packages that are not `HostedDependencies`.

We don't care about git-dependencies and path-dependencies. Those won't exist on pub.dev, but if someone were to analyze their local application then they might have such dependencies.